util-app/src/test/scala/com/twitter/app/AppTest.scala

```diff
@@ -303,6 +311,21 @@ class AppTest extends FunSuite {
     assert(closed)
   }
 
+  test("App: exit even if the ShutdownTimer doesn't schedule timeout promptly") {
+    val app = new TestApp(() => ()) {
+     // We use a MockTimer to simulate jitter in the scheduling of timeout
+     override protected lazy val shutdownTimer = new MockTimer
+     def tick() = shutdownTimer.tick()
+    }
+
+    app.closeOnExit(Closable.make { _ => Future.never })
```

Comment:
Hey, thanks for the patch. I'm not 100% convinced by the premise here and and looking through, but this test is flawed. You have to wrap the `tick()` call in `Time.withCurrentTimeFrozen`, see the other tests here. We also test the close behavior in the presence of a `TimeoutException` with other tests here as well.

Reply:
Thanks for looking into this. I think this simplified example might help to understand:

```scala
val closeDeadline: Time = ???
val closing = Future.never.by(shutdownTimer, closeDeadline, new CloseException)
Await.result(closing, closeDeadline - Time.now) 
```

This snippet could result in a `CloseException` if the timer fulfilled `closing` promptly or a `TimeoutException` otherwise. The `TimeoutException` case, while possible is unlikely since the timeout duration pass to will eventually be converted back to a deadline (by `java.util.concurrent.CountDownLatch`) but using `System.nanoTime` several operation later. However this case is still possible, we observed it in production and currently `App` doesn't handle it properly.

This unit-test doesn't really need to freeze the time, it just need to make sure that the timer doesn't schedule anything before the main function finish. I would argue that using `Time.withCurrentTimeFrozen` would be a little deceptive in that case, since we a looking at  a race-condition involving `Await.result` which rely on `java.util.concurrent.CountDownLatch` for is timeout so the time wouldn't be frozen for this timeout. However I agree that the current unit-test look shady and artificial, but I wasn't sure the best to trigger this race-condition reliably. I modified the unit-test to use a `Timer` that schedule everything 500ms late which is still subject to the same race-condition but way skewed in the other direction and it feel less artificial.

Reply:
Thanks for the information -- my main comment about using `Time.withCurrentTimeFrozen` is that you cannot properly mock time without using it.

But I digress. We would still like to fully understand the issue and the impact of these changes. As you may be aware, `App` is the basis of all of our services and thus there is a very high bar for changes, especially when they deal with the lifecycle of the application, like close -- and definitely when the new logic is wired up to so a `System.exit` -- we would like to thread carefully, here. 

The initial reaction is a strong desire to not introduce a `kill` function in this manner and therefore I think we'd really like to understand the issue and make sure we're actually addressing it and not a side-effect of something else that should or could be better handled elsewhere. 

To that end, your simple example has nothing to do with `App` and thus it is somewhat confusing as to why the solution would be in `App` (at least to me)? My point here is to make sure we are observing and testing the `App` API here if the solution is in `App`, since there is obviously more that happens in `close` than the provided snippet.

To make sure I understand the gist of the problem correctly, you've observed that a `TimeoutException` can escape when closing the application instead of the expected `CloseException` and since the `TimeoutException` is not handled, bad things can happen -- is that accurate? I want to be perfectly clear that we are seeing this behavior through the normal flow of the `App#close` method. Thanks.

Reply:
The `App` currently `Await` on itself at the end of the `main` [ref](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L376). If one of the `Closable` is never fulfilled, this `Await` can result in a `CloseException` or a `TimeoutException` due to a race-condition within the `App`. Currently only the `CloseException` is handled [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L71) and ultimately result in a `System.exit(1)`. The `TimeoutException` would result in calling `exitOnError(reason: String, details: => String)` which will itself throw a `TimeException` [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L96). There's no way to handle that downstream apart from overriding `exitOnError` which doesn't seem like a good idea.

The modified unit-test show the issue and simply call `App#main`. The only "strange" part is the `Timer` that schedule with a delay, but that can happen on the JVM since according to the [doc](https://docs.oracle.com/javase/8/docs/api/java/util/Timer.html) `This class does not offer real-time guarantees: it schedules tasks using the Object.wait(long) method.` If you are still not convinced by this unit-test, [this](https://github.com/twitter/util/compare/develop...gpoulin:simple_timeout?expand=1#diff-9f877f117b521ff3cfd79847c7fcf534R314) simple one fail about half the time when I try it locally, which also mean it succeed about half the time which is not great for a unit-test.

The addition of the `kill` method is only needed for the unit-test, as currently this race condition will cause `exitOnError(reason: String)` to be called (which is what other unit-test check), but this method will throw a `TimeoutException` and never reach `System.exit(1)`. I would be happy to do away with this `kill` method if you know of a way that I can unit-test `exitOnError(reason: String)`.

Reply:
To clarify, by "simplified example", I meant a simplify version of the race-condition that exist in `App`. This example, is basically what `App` does if you remove the logic that doesn't concern this race-condition.

`val closeDeadline: Time = ???` -> [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L208)
`val closing = Future.never.by(shutdownTimer, closeDeadline, new CloseException)` -> [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L280)
`Await.result(closing, closeDeadline - Time.now)` -> [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L376)